Client/Frontend/Shields/BraveShield.swift

```diff
@@ -0,0 +1,173 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveShared
+import Data
+import CoreData
+
+// These override the setting in the prefs
+public enum BraveShield {
+    case AllOff
+    case AdblockAndTp
+    case HTTPSE
+    case SafeBrowsing
+    case FpProtection
+    case NoScript
+    
+    var globalPreference: Bool {
+        switch self {
+        case .AllOff:
+            return false
+        case .AdblockAndTp:
+            return Preferences.Shields.blockAdsAndTracking.value
+        case .HTTPSE:
+            return Preferences.Shields.httpsEverywhere.value
+        case .SafeBrowsing:
+            return Preferences.Shields.blockPhishingAndMalware.value
+        case .FpProtection:
+            return Preferences.Shields.fingerprintingProtection.value
+        case .NoScript:
+            return Preferences.Shields.blockScripts.value
+        }
+    }
+}
+
+private class PrivateBrowsingShieldOverride {
+    fileprivate static var privateModeOverrides: [String: PrivateBrowsingShieldOverride] = [:]
+    
+    private var overrides: [BraveShield: Bool] = [:]
+    
+    init(shield: BraveShield, isOn: Bool) {
+        overrides[shield] = isOn
+    }
+    
+    subscript(shield: BraveShield) -> Bool? {
+        get { return overrides[shield] }
+        set { overrides[shield] = newValue }
+    }
+}
+
+extension Domain {
+    
+    class func setBraveShield(forUrl url: URL, shield: BraveShield, isOn: Bool?,
+                              context: NSManagedObjectContext = DataController.newBackgroundContext()) {
+        if PrivateBrowsingManager.shared.isPrivateBrowsing {
+            setPrivateShieldForDomainUrl(url.domainURL.absoluteString, shield: shield, isOn: isOn, context: context)
+            return
+        }
+        let domain = Domain.getOrCreateForUrl(url, context: context)
+        domain.setBraveShield(shield: shield, isOn: isOn, context: context)
+    }
+    
+    class func getBraveShield(forUrl url: URL, shield: BraveShield,
+                              context: NSManagedObjectContext = DataController.newBackgroundContext()) -> Bool? {
+        if PrivateBrowsingManager.shared.isPrivateBrowsing {
+            return getPrivateShieldForDomainUrl(url.domainURL.absoluteString, shield: shield, context: context)
+        }
+        let domain = Domain.getOrCreateForUrl(url, context: context)
+        return domain.getBraveShield(shield)
+    }
+    
+    func setBraveShield(shield: BraveShield, isOn: Bool?,
+                        context: NSManagedObjectContext = DataController.newBackgroundContext()) {
+        if PrivateBrowsingManager.shared.isPrivateBrowsing {
```

Comment:
Now that https://github.com/brave/brave-ios/pull/621/files#diff-4078b8097af7507d3c5456a747da725dR56
exists, this shouldn't ever be called, since this requires a `Domain` instance.

Can we instead do a full assert here?

```
asset(!PrivateBrowsingManager.shared.isPrivateBrowsing, "Should not be updating Domain objects while in private mode")
```

Reply:
Yeah, actually these instance method versions aren't even used, so I'm going to make them private as well